File: src/objects/lp_model.py
import pyarrow as pa
import pyarrow.compute as pc
from .base_model import ArrowModel
from utils.model_sanity_check import check_arrow_coo_matrix, check_variable_bounds, check_objective_sense

"""
LPModel: Linear Programming Model Representation using PyArrow

This class provides a structured way to represent a standard linear programming (LP) model
using Arrow's in-memory format for high-performance serialization and communication.

Standard LP form:
    minimize     c^T x
    subject to   A x = b
                 lb <= x <= ub

Model Components:
    - A:       Constraint matrix in sparse COO format (as a RecordBatch with columns "row", "col", "val")
    - b:       Right-hand side vector
    - c:       Objective function coefficients
    - lb:      Lower bound for each variable(Optional)
    - ub:      Upper bound for each variable(Optional)
    - osense:  Objective sense, e.g., "min" or "max" (Optional, defaults to "min")
    - csense:  Constraint senses, e.g., ["E", "L", "G"] for equality, ≤, ≥ (Optional, defaults to all "E")
"""

class LPModel(ArrowModel):
    def __init__(
        self,
        A: pa.RecordBatch,
        b: pa.Array,
        c: pa.Array,
        lb: pa.Array = None,
        ub: pa.Array = None,
        osense: pa.Scalar = None,
        csense: pa.Array = None
    ):
        """
    Initialize an LP model.

    Required:
            - A (RecordBatch): Sparse constraint matrix in COO format with "row", "col", "val"
            - b (Array): Right-hand side vector
            - c (Array): Objective function coefficients

        Optional:
            - lb (Array): Lower bounds (default: None, treated as unbounded)
            - ub (Array): Upper bounds (default: None, treated as unbounded)
            - osense (Scalar): "min" or "max" (default: "min")
            - csense (Array): ["E", "L", "G"] (default: all "E")

    Raises:
        TypeError / ValueError if inputs are malformed
    """
        if not isinstance(A, pa.RecordBatch):
            raise TypeError("A must be RecordBatch")
        if not all(name in A.schema.names for name in ["row", "col", "val"]):
            raise ValueError("A must contain 'row', 'col', and 'val' columns")

        self.A = A            # Sparse matrix in COO format (row, col, val)
        self.b = b            # Right-hand side vector
        self.c = c            # Objective function coefficients
        self.lb = lb    # Lower bounds for decision variables
        self.ub = ub    # Upper bounds for decision variables
        self.osense = osense if osense is not None else pa.scalar("min", type=pa.string())
        self.csense = csense if csense is not None else pa.array(["E"] * len(b), type=pa.string())
              
              
              
    def sanity_check(self):
        """
        Perform consistency checks on LP model dimensions and indices.

        This includes:
            - Shape consistency between variable vectors and bounds
            - Index validity in sparse matrix A
            - Constraint sense and objective sense validity

        Raises:
            ValueError: If any mismatch or invalid structure is detected.
        """
        n_vars = len(self.c)
        n_cons = len(self.b)

        # 1. Check sparse matrix A: row/col index validity
        check_arrow_coo_matrix("A", self.A, n_cons, n_vars)

        # 2. Check bounds (optional)
        check_variable_bounds(self.lb, self.ub, n_vars)

        # 3. Check csense (must match number of constraints)
        if len(self.csense) != n_cons:
            raise ValueError(f"Length of csense ({len(self.csense)}) != number of constraints ({n_cons})")
        for i, val in enumerate(self.csense):
            if val.as_py() not in {"E", "L", "G"}:
                raise ValueError(f"Invalid csense[{i}] = {val.as_py()} (must be one of 'E', 'L', 'G')")

        # 4. Check osense
        check_objective_sense(self.osense)
        
        
    # static method to create an LPModel from a dictionary representation
    @classmethod
    def from_dict(cls, model_dict: dict) -> "LPModel":
        """
        Create an LPModel instance from a dictionary representation.

        :param model_dict: Dictionary with keys 'A', 'b', 'c', 'lb', 'ub', 'osense', 'csense'
        :return: LPModel instance
        """
        A = pa.RecordBatch.from_pydict(model_dict["A"])
        b = pa.array(model_dict["b"])
        c = pa.array(model_dict["c"])
        lb = pa.array(model_dict.get("lb", []))
        ub = pa.array(model_dict.get("ub", []))
        osense = pa.scalar(model_dict.get("osense", "min"), type=pa.string())
        csense = pa.array(model_dict.get("csense", ["E"] * len(b)), type=pa.string())
        
        return cls(A, b, c, lb, ub, osense, csense)

File: src/service/optimization_service/julia/model/lp_model.jl
"""
LPModel module provides functionality to create and manage linear programming models using JuMP.
It includes functions to build a JuMP model from problem data and to set up the optimization problem.
"""
module LPModel

using SparseArrays
using JuMP
include("../utils/sparse_matrix.jl")

export build_jump_model

"""
    build_jump_model(data::Dict)
Builds a JuMP model for linear programming based on the provided data dictionary.

# Arguments
- `data::Dict`: A dictionary containing the optimization problem data, including:
  - `:A`: Arrow table representing the constraint matrix.
  - `:b`: Right-hand side vector.
  - `:c`: Objective function coefficients.
  - `:lb`: Lower bounds for the decision variables.
  - `:ub`: Upper bounds for the decision variables.
  - `:csense`: Constraint senses (e.g., "E", "G", "L").
  - `:osense`: Objective sense ("min" or "max").

# Returns
- `JuMP.Model`: A JuMP model instance configured for linear programming.
- `JuMP.Variable`: The decision variables of the model.
- `JuMP.Expression`: The objective function expression of the model.

# Workflow
1. Extracts and converts the data from the dictionary.
2. Constructs the constraint matrix `A` using sparse representation.
3. Sets up the JuMP model with the objective function and constraints.

# Example
```julia
data = Dict(
    :A => Arrow.Table(row=[0, 1], col=[0, 1], val=[1.0, 2.0]),
    :b => [3.0, 4.0],
    :c => [1.0, 2.0],
    :lb => [0.0, 0.0],
    :ub => [10.0, 10.0],
    :csense => ["E", "G"],
    :osense => "min"
)
model, x, c = build_jump_model(data)
```
"""
function build_jump_model(data)
    A_data = data[:A]
    b = Vector{Float64}(data[:b])
    c = Vector{Float64}(data[:c])
    lb = Vector{Float64}(data[:lb])
    ub = Vector{Float64}(data[:ub])
    csense_strs = Vector{String}(data[:csense])
    osense_str = data[:osense]  # e.g. "max"
    osense = osense_str == "max" ? -1 : 1  # 1 for min which is JuMP default, -1 for max

    A = to_sparse(A_data)
    # sense_map = Dict("E" => '=', "G" => '≥', "L" => '≤')
    sense_map = Dict("E" => '=', "G" => '>', "L" => '<')
    csense = [sense_map[c] for c in csense_strs]

    return buildlp(c * osense, A, csense, b, lb, ub)
end


"""
    buildlp(c, A, sense, b, l, u)
Builds a linear programming model using JuMP.

# Arguments
- `c`: Coefficients of the objective function.
- `A`: Constraint matrix in sparse format.
- `sense`: Constraint senses (e.g., '=', '>', '<').
- `b`: Right-hand side vector for the constraints.
- `l`: Lower bounds for the decision variables.
- `u`: Upper bounds for the decision variables.

# Returns
- `JuMP.Model`: The JuMP model instance.
- `JuMP.Variable`: The decision variables of the model.
- `JuMP.Expression`: The objective function expression of the model.

# Workflow
1. Initializes a JuMP model.
2. Defines decision variables with bounds.
3. Sets the objective function to minimize the linear expression.
4. Adds constraints based on the provided matrix and senses.
5. Returns the model, decision variables, and objective coefficients.

# Example
```julia
c = [1.0, 2.0]
A = sparse([1.0, 2.0], [1.0, 2.0], [1.0, 2.0], 2, 2)
sense = ['=', '≥']
b = [3.0, 4.0]
l = [0.0, 0.0]
u = [10.0, 10.0]
model, x, c = buildlp(c, A, sense, b, l, u)
```
"""
function buildlp(c, A, sense, b, l, u)
    N = length(c)
    model = Model()
    x = @variable(model, l[i] <= x[i=1:N] <= u[i])
    @objective(model, Min, c' * x)
    eq_rows, ge_rows, le_rows = sense .== '=', sense .== '>', sense .== '<'
    @constraint(model, A[eq_rows, :] * x .== b[eq_rows])
    @constraint(model, A[ge_rows, :] * x .>= b[ge_rows])
    @constraint(model, A[le_rows, :] * x .<= b[le_rows])
    return model, x, c
end

end # module LPModel